### ISSUE TITLE
Make it easier to configure and understand can-set

### ISSUE THREAD

**Issue by chasenlehara** [canjs/can-set#54]
__tldr;__ It's difficult to make `can-set` work with a custom service endpoint. We propose a new API to make it more straightforward.

> This was discussed at a [contributors meeting (35:26)](https://youtu.be/95OX6-qii24?t=35m26s)

## The Problems

There are two core problems we'd like to address with `can-set`:

- It's _API_ is aligned with set theory instead of terms users are more familiar with.
- It's too difficult to configure a custom service endpoint.

#### API Naming

`can-set` relies heavily on set theory.  However, it's not really something the creator of a set algebra needs to understand.  Instead, they are modeling their _get many_ service parameters. Furthermore, where users are not familiar with set theory, users are typically familiar with _databases_ and _sql_ terminology. 

#### Custom Service Endpoints

It's difficult to make `can-set` work with a custom service endpoint.  For example, lets say your service layer looked like:

```
todos?slice[start]=0&slice[end]=100
```

`can-set` currently expects people to use _set objects_ that match the query parameters. So the _set object_ for the above code looks like:

```js
{slice: {first: 0, last: 100} }
```

By default, `can-set` doesn't support pagination, so this needs to be added into the algebra:

```js
algebra = new set.Algebra(
  set.props.rangeInclusive("first","last")
)
```

Also, `first` and `last` are inside a `paginate` object, so this needs to be added:

```js
algebra = new set.Algebra(
  set.props.rangeInclusive("first","last"),
  new set.Translate("paginate","slice")
)
```

In order to keep _set objects_ looking like the query parameters, `can-set` requires users to perform a lot of complex configuration.

## Solution Proposal - can-service and a standard `query` type

This proposal involves two parts:

1.  A standard `query` type that will be used to perform set logic.  All `can-connect` `.getList(query)`, `.getListData(query)`, and similar methods will use this type.  No more custom `set` types. `can-query` would be an implementation of `can-set` that would ONLY work on these `query` data types.
1. A `can-service` constructor that is used to describe the relationship between the standard `query` type and the parameters used on the service layer. This will be used by `can-connect` and `can-fixture`.  

### can-query

The most significant part of this proposal is that we'd make it so people MUST use a standard `query` type when using `can-connect`.

For example, currently it is possible to configure an algebra so the following parameters to `getList` works:

```js
Todo.getList( {slice: {first: 0, last: 100} } )
```

We propose making it so `.getList()` must take the following in order for it to have an understanding of pagination:

```js
Todo.getList({ paginate: {  start: 0,  end: 100 } })
```

#### Query Type

I propose the following for the standard `query` type:

```js
{
  // where: AND_CLAUSE|OR_CLAUSE
  where: {
    typeID: 21,
    name: { $ne: null },
  },
  paginate: {
    skip: 10,
    limit: 19
  },
  sort: { 
    name: 1,
    age: -1
  },
  // RESERVED
  select: ["property1", "property2.property3"],
  related: ["owners"],
  group: {age: "count"}
}
```

AND_CLAUSE: 

```js
//Object<(propertyName,value|QUALIFIER) | ($or,OR)>
{ taskId: 1 }
```

OR_CLAUSE: 

```js
// Array<AND_CLAUSE>
[
	{roomId: 2},
	{
		archived: true,
		$or: [
			{foo: "bar"},
			{a: "b"}
		]
	}
]
```

QUALIFIER:

```js
// The value is one of the items in the array
// $in: ARRAY<value>
{ $in : ['red', 'blue'] }

// The value is not one of the items in the array
{ $nin: [4, 6] }

// Compare to a value
{ $lt: 6 }
{ $lte: 6 }
{ $gt: 6 }
{ $gte: 6 }
{ $ne: null }
```

This query type is a slightly modified [MongoDB query language](https://docs.mongodb.com/manual/reference/operator/query-comparison/).  The only difference is that 
all clauses (notably the where and paginate clause) are properties of the _query_ type.  In my opinion, it's worth breaking form this well known API for several reasons:

1. Implementor Ease - It's easy to identify which parts of the query `can-query` MUST operate on (where, paginate, sort) and which parts it won't (everything else).
1. User clarity - It forces the user to group related information.

#### Benefits of a standard query interface

A standard query interface allows the construction of widgets that can work with 
unknown connection types.  For example, a sortable and paginate-able grid can be created because to get 20 items sorted by `name`, one will always write:

```js
connection.getList({sort: {name: 1}, paginate: {skip: 0, limit: 20}})'
``` 

#### `can-query` Use

`can-query` would have the same API as [can-set.Algebra](https://canjs.com/doc/can-set.Algebra.html)'s methods:

```js
var query = require("can-query");

query.subset( {where: {todoId: 1}}, {where: {}}) //-> true
```

### can-service

`can-service` is use to describe a service endpoint including how it relates to the `query` type.  It is very similar to `can-connect/data/url` and `can-connect/data/parse`.  However, it will also be useful to can-fixture because it can translate a request to the `query` type.

```js
var Service = require("can-service");
var service = new Service({
  // only necessary to specify `id` properties if not already `id`.
  // additional properties could be used by `can-fixture` to generate random data
  schema: {
    id: "id"
  }, 
  // Given `query`, make a request for the data
  getListData(query){
    return Promise || REQUEST
  },
  // Converts `REQUEST` back into a query.  Useful for fixtures.
  getListQuery(REQUEST){
    return query
  },

  getData(idQuery){  },
  getIdQuery(){  }, // not a fan of this name getOneQuery?

  createData(instanceData) {  return PROMISE | REQUEST },
  getCreatedData(REQUEST){  return instanceData },

  updateData(instanceData){ return PROMISE | REQUEST },
  getUpdateData(REQUEST){ return instanceData },

  patchData(patchData),
  getPatchData(REQUEST)

  deleteData(deleteData)
  getDeleteData(REQUEST)

  // Should we add methods for converting data into what the service produces?
  toServiceListData(listData),
  toServiceInstanceData(instanceData),
  toServiceCreateData( instanceData ),
  toServiceUpdateData( instanceData ),
  toServicePatchData( patchData ),
  toServiceDeleteData( deleteData )
});
```

##### Example

Lets see what it might look like to configure a basic _get many_ restful service layer where 
requesting the second page of 10 people who are `typeID=21` and have a name would look like:

```js
peopleConnection.getList({
  where: {
    typeID: 21,
    name: { $ne: null }
  },
  paginate: {
    skip: 10,
    limit: 10
  }
})
```

And the request would need to be made like:

 `/people?typeID=21&name=notnull&start=10&end=19`

> NOTE: the where and paginate clause are not nested and it uses start and end.

The service might be configured like:

```js
var Service = require("can-service");
var personService = new Service({
  schema: {id: "id"}
  getListData: function(query){
    var request = { url: "/people", method: "get", data: {} };
    
    Object.assign(request.data, query.where);
    request.data.start = query.paginate.skip;
    request.data.end = query.paginate.skip + query.paginate.limit - 1;
    return request; // or ajax(request);
  },
  getListQuery: function(request){
    var query = {where: {}, paginate: {}};
    
    query.paginate.limit = request.data.end - request.data.start + 1;
    query.paginate.skip = request.data.start;

    delete request.data.start;
    delete request.data.end;
    Object.assign( query.where, request.data);
  }
})
```

The people service would be provided to the `peopleConnection`:

```js
connect([ ... ], {
  service: peopleService
})
```

And it can be used to create fixtures:

```js
var store = fixture.store( [/* items in store */ ] );
fixture( peopleService, store );
```

While `can-service` provides a constructor useful for a wide variety of service endpoints, there will be other modules that make configuration easier. For example:

```js
var RestService = require("can-service-rest");
new RestService({
  url: "/people",
  toData: function(query){
    var data = {};
    Object.assign(data, query.where);
    data.start = query.paginate.skip;
    data.end = query.paginate.skip + query.paginate.limit - 1;
    return data; 
  },
  toQuery: function(data){
     var query = {where: {}, paginate: {}};
    
    query.paginate.limit = data.end - data.start + 1;
    query.paginate.skip = data.start;

    delete data.start;
    delete data.end;
    Object.assign( query.where, request.data);
  }
});

var FeathersService = require("can-service/feathers");
new FeathersService({
  url : "/people"
});
```

## Tasks

- [x] Determine generic `query` structure.
- [ ] Implement something like an algebra with this.
- [ ] Figure out how to support mixins
- [ ] Update `can-connect` to support this
- [ ] Update CanJS's guides
- [ ] Update DoneJS

**Comment by marshallswain** [comment]
An idea for the API: https://gist.github.com/marshallswain/7c70b9049ecd3b7fe58d7e34c40f9034

```js
var query = {
  _id: '208hvl2l9ehbecoi3o',
  lastName: 'Fadeev',
  retired: true,
  $sort: {
    firstName: 1
  },
  $skip: 5,
  $limit: 10,
  $select: ['field1', 'field2'],
  favoriteColor: {
    $in: ['red', 'blue'],
  },
  numberOfChildren: {
    $nin: [4, 6],
  },
  dateBoughtFirstCar: {
    $lt: new Date('2010-01-05'),
  },
  numberOfPaintingsOwned: {
    $lte: 6,
  },
  createdAt: {
    $gt: 0,
  },
  lastSeen: {
    $gte: 1,
  },
  archived: {
    $ne: null
  },
  $or: [
    { archived: { $ne: true } },
    { roomId: 2 }
  ]
}
```

**Comment by nlundquist** [comment]
I like the client supported query format that @marshallswain has proposed and wonder if it is equivalent to an existing non-SQL DB query format? Looks very much like Mongo. If there is an existing "query language" we could base this off of I think that would be the best approach. 

We should make a point of telling users they can avoid writing a custom `can-service` description altogether if they support this query format on their service layer. Many users with existing APIs wouldn't be able to avoid customization, but it's worth noting for those that could.

I think that `toServiceFormat` and `toClientFormat` or similar would be more descriptive than `toParams` and `toQuery`. It's very common to refer to the "query parameters" as the data sent to the service and it would very quickly become unclear if a "query" meant the data provided to the service or the data provided to a can-connect method. Both sets of data are arguably queries.

We'd probably also need to support making `id` an array or function to allow for compound ids, similar to how can-set currently allows multiple id properties.

The conversion API may need to do more than provide a set of parameters to be converted into a query string. IMO since it's essentially a "request middleware" it should be able to modify all parts of a potential request. Though atypical there are APIs that expect a GET request to provide a request body containing the query parameters rather than using a query string, I'm thinking of the Elasticsearch API specifically. Aside from query string and request body, the API should also allow client query parameters to be converted into headers in the service request. Maybe provide a simple return format and a more complicated one, something like:
```
  // basic request format
  toServiceFormat: function(query){
    ...
    return {limit: 10, offset: 20, ...}; // turns this whole object into a query string
  }
  // and an alternative return format 
  toServiceFormat: function(query){
    ...
    // can provide query data in multiple parts of the request
    return {
      querystring: {limit: 10, offset: 20}, 
      headers: {'x-select-shard': 'archive'},
      requestbody: "<user>"+10+"</user>"
    }; 
  },
```
Users might also have a need to convert client query parameters into part of the request URL itself for service endpoints like `/student/10/exam_result/midterm/`. 

Granted these examples aren't very good API design, but there is a lot of that out there and users would likely expect a flexible solution to be able to cope with that. The REST conventions are all too often not followed well and we'd need to decide if `can-service` would require them to be.

**Comment by justinbmeyer** [comment]
@nlundquist The main problem I think exists w/ @marshallswain 's syntax proposal is it doesn't separate `where` clause from everything else.  Imo, it's much easier in terms of implementation to have the different types of data transformation clauses separated.

**Comment by marshallswain** [comment]
I think it would be smart to align (with zero-configuration) with a query language that's widely used.  There are only a few options that fit that description.  

- GraphQL - which is the most popular because of Facebook Engineering.  Unfortunately, it lacks support for realtime, so no built in WebSockets or Server-Side Events.  Database support is limited to a handful of options, but it still wins the gold medal for the most hype.
- FeathersJS - uses the exact same query language as MongoDB, and it's what I built my examples around.  It aligns nicely with what the Meteor devs are accustomed to using.  The [list of supported databases is the largest around.](https://docs.feathersjs.com/ecosystem/readme.html#database), and all of those databases are supported with the exact same query language (even without separating the `where` clauses).
- Waterline, from SailsJS - is pretty similar to the MongoDB syntax, supports many databases, but not nearly as many as Feathers.

> it's much easier in terms of implementation to have the different types of data transformation clauses separated.

We haven't found this to be a problem with any of the database adapters we've written for FeathersJS.  Even if it were a real problem, the accuracy of this statement depends on which level of "implementation".  It might be easier for the ones writing can-set's code, but any custom syntax we support will only create more friction for those who want to implement can-set as part of their project.  I'd love to see a zero-friction implementation working with a backend that also allows customization to work with other syntax.  Obviously, my vote would be to support MongoDB/FeathersJS out of the box and to enable support for custom syntax, like separating a `where` clause, through things that amount to additional project overhead.

Again, I'm super biased, but I really want this to work with FeathersJS out of the box, with no effort.  Then I gain support for MySQL, Postgres, MongoDB, RethinkDB, Solr, ElasticSearch, etc., without having to learn another query language.  But I think my bias is a short stretch.

**Comment by justinbmeyer** [comment]
Having experienced this problem already with can-set, which caused the introduction of Translate, I'd like to avoid it. We can give people nice warnings if there is an unsupported clause. We won't be able to do that if every where is directly on the query object. 

Also, while zero configuration is nice, can-set isn't only a translation to some sql or other API. It's mostly about comparing sets. This isn't something feathers does afaik. This is an important difference. Separating out the clauses was something I had to hack into can-set. I would like to remove this code. 

I think doing something like service(feathers) or service(abc) isn't so bad. In the end, it's the same except "where" is clearly identified.

**Comment by justinbmeyer** [comment]
After thinking about this more, here's the plan as far as I can see it.

First, I discovered that a standard query language on its own isn't enough.  We need a schema/type system.  For example, how does the following query know what all the possible values of `color` are without an enum or set type to tell it:

```js
{
  where: {color: ["red"]}
}
```

As I'd prefer users not have to duplicate schema information (one for their Observable type, another for their query), I'd like the schema to be read-able from `DefineMap`, `can-observe`, etc.

For example, something like this should work:

```js
Todo = DefineMap.extend({
  id: {identity: true, type: "number"},
  name: "string",
  complete: {type: "boolean"},
  status: {type: Enum(["new","assigned","complete"])
})

new RestConnection({
  url: "/todos",
  InstanceType: Todo
})
```

Internally, `RestConnection` will create a query language using:

```js
var todoQuery = new Query(Todo);
```

This will be able to understand queries like:

```js
{
  where: {color: ["red"]}
}
```

Therefore, I'd like `can-reflect` or a library "just" ontop of `can-reflect` to enable type specifications.  (_this will also be useful for [can-crud style prototyping](https://github.com/donejs/donejs/issues/1050)_).

I'm currently calling this `can-type`.  Given a constructor function like `Todo` it should be able to return a "schema" for it like:

```js
Type.schema(Todo) //-> {
  type: "map",
  keys: [
    {key: "todo", type: String}
  ]
}
```

I'm leaning away from `canReflect.getInstanceKeyDescriptor(Type, key)` in the [quick prototype](https://github.com/donejs/donejs/issues/1050) issue.  This is because I would like to support non-object types more easily, like `real-number-range`:

```js
QUnit.equal(
    Type.isSubset(
        new RealNumberRangeInclusive(1,4),
        new RealNumberRangeInclusive(0,5)
    ),
    true,
    "1-4 subset of 0-5");
```

**Comment by matthewp** [comment]
This became can-query-logic, right? So can we close this issue?
